### PR TITLE
JENKINS-36330 made it possible to set max number of jobs to display in a view

### PR DESCRIPTION
**Pull request by bobo**
Tried it localy with hpi:run and it seems to work fine.
Added some new tests but not sure if i missed any kind of test?

Not sure maxNumberOfJobs is a great name but cant think of anything better, suggestions welcomme.

### CODE REVIEW

**Comment by tommysdk** [comment]
@bobo I want to make sure I fully understand the intent of this feature request. Why is it not sufficient to sort on latest activity and make sure only the pipelines of interest are shown?

I'm not convinced we should hide pipelines that are configured to be visible by the current view. Rather I would suggest just configuring the view accordingly, e.g. through JobDSL to display the pipelines of interest. Would that be a sufficient workaround for this use case?

Could you please elaborate your use case and solution further to make sure we fully understand your needs.

**Comment by bobo** [comment]
@tommysdk what i want is to be able to have the most interesting pipelines (ie most recently used) on one monitor.  And while this is possible without hiding them, it gets ugly when you only see the headline of a pipeline for example. So to be clear, its only for cosmetic reasons.
I do however agree that its bad that some gets hidden and you dont know. Maybe there should be something in the headline of the view? For ex "Displaying 6/34 pipelines"

For a case you can look on the monitor behind the tv in our area at work :-) or the /most-recent view

**Comment by tommysdk** [comment]
@bobo I think this feature would only make sense if using the sort by latest activity. Maybe we could provide this option only when latest activity sorting has been enabled (default option being all). WDYT?

**Comment by bobo** [comment]
@tommysdk i think that makes sense.
Il look into how to solve that, and also see how it looks with a "showing 5/24 pipelines" thing

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly [hunk 1]
```diff
@@ -27,6 +27,10 @@
             <f:select/>
         </f:entry>
 
+        <f:entry title="Max number of pipelines" field="maxNumberOfJobs" default="-1">
+            <f:textbox/>
+        </f:entry>
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly]: Is it possible to only show this in the UI if sort by latest activity is selected?

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/help-maxNumberOfJobs.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Max number of Jobs to display on the page.
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/help-maxNumberOfJobs.html]: I suggest updating to a more descriptive text, something along the lines with: 
"Limits the number of pipelines shown in the view, regardless of how many pipelines are configured"

File: src/test/java/se/diabol/jenkins/pipeline/DeliveryPipelineViewTest.java [hunk 3]
```diff
@@ -387,6 +391,59 @@ public void testGetPipelineViewWithLastJobProvided() throws Exception {
...
+        view.setMaxNumberOfJobs(2);
+        assertEquals(regExpSpecs, view.getRegexpFirstJobs());
+
+        jenkins.getInstance().addView(view);
+
+        List<Component> components = view.getPipelines();
+        assertNull(view.getError());
+        assertEquals(2, components.size());
+    }
+
```
Comment (tommysdk) [inline src/test/java/se/diabol/jenkins/pipeline/DeliveryPipelineViewTest.java]: Good test cases!

File: src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java [hunk 4]
```diff
@@ -111,7 +111,7 @@
     private boolean showAggregatedChanges = false;
     private String aggregatedChangesGroupingPattern = null;
     private String theme = DEFAULT_THEME;
-
+    private int maxNumberOfJobs = -1;
```
Comment (tommysdk) [inline src/main/java/se/diabol/jenkins/pipeline/DeliveryPipelineView.java]: Maybe we should give it a more descriptive name, such as "maxNumberOfVisiblePipelines" (and update the name of the test cases accordingly)

**Comment by tommysdk** [comment]
I've tested the solution and I think it looks good. I think we could incorporate it after my recent code comments have been addressed.

Would be nice to try it out with a "showing 5 of 24 pipelines" text in the views as well.

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly (hunk 1 above)
Comment (patbos) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly]: @tommysdk Dont see the reason why

Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly]: @patbos I don't think limiting the number of pipelines makes sense without the sorting. If you have a static order why not just remove the pipelines that are not shown from the pipeline view configuration? But I'm open to either solution.

Comment (patbos) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly]: @tommysdk Then we should think of refactor and rename the ComponentComparator class and make an option for it to limit the result.

Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/configure-entries.jelly]: Let's go a head with the current solution with the below code comments addressed. I think the proposed solution is sufficient for this use case.

**Comment by tommysdk** [comment]
Addressed code comments and merged manually: 2f01d75ba12f8581c9bbdf8037387ed87894c390

**Comment by tommysdk** [comment]
Merge successfully passed CI: https://travis-ci.org/Diabol/delivery-pipeline-plugin/builds/156631327

Thank you @bobo for your contribution to the Delivery Pipeline Plugin!

**Pull request closed without merging**